packages/realm-react/src/useAuth.tsx

```diff
@@ -0,0 +1,196 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2023 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+import { useCallback } from "react";
+import { useApp, useAuthResult } from "./AppProvider";
+import { AuthError, AuthOperationName, AuthResult, OperationState } from "./types";
+import { Realm } from "realm";
+import { useAuthOperation } from "./useAuthOperation";
+
+/**
+ * Hook providing operations and corresponding state for authenticating with a
+ * Realm app.
+ *
+ * The {@link AuthResult} states returned from this hook are "global" for all
+ * components under a given AppProvider, as only one operation can be in progress
+ * at a given time (i.e. we will store the states on the context). This means that,
+ * for example, multiple components can use the `useAuth` hook to access
+ * `loginResult.pending` to render a spinner when login is in progress, without
+ * needing to pass that state around or store it somewhere global in their app
+ * code.
+ */
+interface UseAuth {
+  /**
+   * Log in with a {@link Realm.Credentials} instance. This allows login with any
+   * authentication mechanism supported by Realm.
+   * If this is called when a user is currently logged in, it will switch the user.
+   *
+   * @returns A `Realm.User` instance for the logged in user.
+   */
+  logIn(credentials: Realm.Credentials): Promise<Realm.User | void>;
```

Comment:
In which was will this (or any of the other methods) return a promise of `void`? 🤔

Reply:
If it fails

Reply:
That should reject the promise? Not resolve it with `undefined`?

Reply:
Then would the correct type be `Promise<Resul.User | AuthError>`?

Reply:
Nope. Just `Promise<Result.User>` there's no generic type for the value of the rejection.

Reply:
I'll do some reviewing here on how promises should reject in the error case.  Throwing is unsafe in a hooks environment.

Reply:
Now everything returns void :100: